### PR TITLE
Improvements in command handling #2

### PR DESCRIPTION
**Pull request by TalZaccai**
This PR is based on #476 

- [x] Eliminate re-parsing of command tokens by accessing parseState in RespServerSession
  - [x] BasicCommands
  - [x] AdminCommands
  - [x] CustomRespCommands
  - [x] InfoCommand
  - [x] RespLatencyCommand
  - [x] ArrayCommands
  - [x] BitmapCommands
  - [x] HyperLogLogCommands
  - [x] KeyAdminCommands
  - [x] HashCommands
  - [x] ListCommands
  - [x] SetCommands
  - [x] SharedObjectCommands
  - [x] SortedSetCommands
  - [x] SortedSetGeoCommands
  - [x] PubSubCommands
  - [x] TxnRespCommands
- [x] Refactor ProcessAdminCommands

To-do in a consecutive PR:
- Introduce ObjectInput struct to simplify object command execution
- Eliminate re-parsing of command tokens in the object implementation
- Relay parseState to transaction / custom command execution (eliminate GetNextArg)

**Benchmark results:**
**main:**
| Method     | Job    | EnvironmentVariables | Runtime  | Mean      | Error     | StdDev    | Allocated |
|----------- |------- |--------------------- |--------- |----------:|----------:|----------:|----------:|
| InlinePing | .NET 6 | Empty                | .NET 6.0 |  1.425 us | 0.0247 us | 0.0231 us |         - |
| Set        | .NET 6 | Empty                | .NET 6.0 | 13.283 us | 0.1302 us | 0.1218 us |         - |
| SetEx      | .NET 6 | Empty                | .NET 6.0 | 18.150 us | 0.1187 us | 0.1110 us |         - |
| Get        | .NET 6 | Empty                | .NET 6.0 |  7.601 us | 0.0404 us | 0.0378 us |         - |
| ZAddRem    | .NET 6 | Empty                | .NET 6.0 | 76.260 us | 0.8081 us | 0.7559 us |   23552 B |
| LPushPop   | .NET 6 | Empty                | .NET 6.0 |  6.900 us | 0.0280 us | 0.0262 us |    1208 B |
| SAddRem    | .NET 6 | Empty                | .NET 6.0 | 63.656 us | 0.4378 us | 0.3881 us |   16384 B |
| HSetDel    | .NET 6 | Empty                | .NET 6.0 | 82.640 us | 1.0457 us | 0.9270 us |   55297 B |
| InlinePing | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  1.346 us | 0.0197 us | 0.0312 us |         - |
| Set        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  9.172 us | 0.0518 us | 0.0459 us |         - |
| SetEx      | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 13.215 us | 0.2607 us | 0.2677 us |         - |
| Get        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  6.277 us | 0.0388 us | 0.0344 us |         - |
| ZAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 62.530 us | 0.5443 us | 0.5091 us |   23552 B |
| LPushPop   | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  6.425 us | 0.0356 us | 0.0316 us |    1208 B |
| SAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 51.035 us | 0.4144 us | 0.3876 us |   16384 B |
| HSetDel    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 69.773 us | 1.0519 us | 0.9839 us |   55297 B |

**Current branch:**
| Method     | Job    | EnvironmentVariables | Runtime  | Mean      | Error     | StdDev    | Allocated |
|----------- |------- |--------------------- |--------- |----------:|----------:|----------:|----------:|
| InlinePing | .NET 6 | Empty                | .NET 6.0 |  1.455 us | 0.0222 us | 0.0208 us |         - |
| Set        | .NET 6 | Empty                | .NET 6.0 | 13.297 us | 0.1195 us | 0.1118 us |         - |
| SetEx      | .NET 6 | Empty                | .NET 6.0 | 17.997 us | 0.1040 us | 0.0973 us |         - |
| Get        | .NET 6 | Empty                | .NET 6.0 |  7.542 us | 0.0452 us | 0.0423 us |         - |
| ZAddRem    | .NET 6 | Empty                | .NET 6.0 | 75.908 us | 1.1988 us | 1.0010 us |   23552 B |
| LPushPop   | .NET 6 | Empty                | .NET 6.0 |  6.946 us | 0.0703 us | 0.0587 us |    1208 B |
| SAddRem    | .NET 6 | Empty                | .NET 6.0 | 62.713 us | 0.8867 us | 0.7861 us |   16384 B |
| HSetDel    | .NET 6 | Empty                | .NET 6.0 | 84.752 us | 1.2541 us | 1.1117 us |   55297 B |
| InlinePing | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  1.381 us | 0.0112 us | 0.0105 us |         - |
| Set        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  9.114 us | 0.1059 us | 0.0990 us |         - |
| SetEx      | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 12.934 us | 0.1139 us | 0.0889 us |         - |
| Get        | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  5.604 us | 0.0530 us | 0.0495 us |         - |
| ZAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 61.009 us | 0.4549 us | 0.4255 us |   23552 B |
| LPushPop   | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 |  6.448 us | 0.0320 us | 0.0299 us |    1208 B |
| SAddRem    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 50.481 us | 0.4520 us | 0.4228 us |   16384 B |
| HSetDel    | .NET 8 | DOTNET_TieredPGO=0   | .NET 8.0 | 67.759 us | 0.7687 us | 0.6815 us |   55297 B |

### CODE REVIEW

File: libs/server/Resp/ArrayCommands.cs [hunk 1]
```diff
@@ -269,15 +269,21 @@ private bool NetworkDEL<TGarnetApi>(ref TGarnetApi storageApi)
...
         /// </summary>
-        /// <param name="ptr"></param>
         /// <returns></returns>
-        private bool NetworkSELECT(byte* ptr)
+        private bool NetworkSELECT()
         {
-            // Read index
-            if (!RespReadUtils.ReadIntWithLengthHeader(out var result, ref ptr, recvBufferPtr + bytesRead))
-                return false;
+            if (parseState.count != 1)
```
Comment (yrajas) [inline libs/server/Resp/ArrayCommands.cs:275]: How about checking these counts based on arity in a common way across all commands?

File: libs/server/Resp/Bitmap/BitmapCommands.cs [hunk 2]
```diff
@@ -113,39 +113,30 @@ internal sealed unsafe partial class RespServerSession : ServerSessionBase
         /// The bit is either set or cleared depending on value, which can be either 0 or 1.
         /// When key does not exist, a new key is created.The key is grown to make sure it can hold a bit at offset.
         /// </summary>
-        private bool NetworkStringSetBit<TGarnetApi>(byte* ptr, ref TGarnetApi storageApi)
+        private bool NetworkStringSetBit<TGarnetApi>(ref TGarnetApi storageApi)
```
Comment (badrishc) [inline libs/server/Resp/Bitmap/BitmapCommands.cs:116]: Not for this PR, but wanted to highlight that none of these methods need to be `bool` any longer, correct?

**Review by badrishc — APPROVED** [review]

File: libs/server/Resp/Bitmap/BitmapCommands.cs (hunk 2 above)
Comment (TalZaccai) [reply libs/server/Resp/Bitmap/BitmapCommands.cs:116]: That's correct

File: libs/server/Resp/ArrayCommands.cs (hunk 1 above)
Comment (TalZaccai) [reply libs/server/Resp/ArrayCommands.cs:275]: Yup, that's on the to-do list

File: libs/server/Resp/Bitmap/BitmapCommands.cs (hunk 2 above)
Comment (badrishc) [reply libs/server/Resp/Bitmap/BitmapCommands.cs:116]: OK, let us add this to a future cleanup.

File: libs/server/Resp/Parser/SessionParseState.cs [hunk 3]
```diff
@@ -110,5 +110,37 @@ public bool Read(int i, ref byte* ptr, byte* end)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int GetInt(int i)
             => ParseUtils.ReadInt(ref Unsafe.AsRef<ArgSlice>(bufferPtr + i));
+
+        /// <summary>
+        /// Try to get int argument at the given index
+        /// </summary>
+        /// <returns>True if integer parsed successfully</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryGetInt(int i, out int value)
```
Comment (yrajas) [inline libs/server/Resp/Parser/SessionParseState.cs:119]: Could be helpful to identify bugs to have an assert that the index is within count

File: libs/server/Resp/Objects/HashCommands.cs [hunk 4]
```diff
@@ -31,66 +31,65 @@ private unsafe bool HashSet<TGarnetApi>(RespCommand command, int count, byte* pt
...
-                if (NetworkSingleKeySlotVerify(key, false))
-                {
-                    return true;
-                }
+            var sbKey = parseState.GetArgSliceByRef(0).SpanByte;
+            var keyBytes = sbKey.ToByteArray();
 
-                // Prepare input
-                var inputPtr = (ObjectInputHeader*)(ptr - sizeof(ObjectInputHeader));
+            var ptr = sbKey.ToPointer() + sbKey.Length + 2;
```
Comment (yrajas) [inline libs/server/Resp/Objects/HashCommands.cs:38]: Nit: A lot of places repeat this pattern ToPointer + Length + 2. Could be extracted into a small inline/extension method.

File: libs/server/Resp/KeyAdminCommands.cs [hunk 5]
```diff
@@ -135,26 +90,24 @@ private bool NetworkGETDEL<TGarnetApi>(byte* ptr, ref TGarnetApi garnetApi)
...
         /// <typeparam name="TGarnetApi"></typeparam>
         /// <param name="count"></param>
-        /// <param name="ptr"></param>
         /// <param name="storageApi"></param>
         /// <returns></returns>
-        private bool NetworkEXISTS<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
+        private bool NetworkEXISTS<TGarnetApi>(int count, ref TGarnetApi storageApi)
             where TGarnetApi : IGarnetApi
         {
+            if (count < 1)
```
Comment (yrajas) [inline libs/server/Resp/KeyAdminCommands.cs:98]: EXISTS supports multiple keys... count could be greater than 1.

**Review by yrajas — APPROVED** [review]

**Pull request merged**